netbench-orchestrator/src/russula/mod.rs

```diff
@@ -0,0 +1,177 @@
+    fn [< is_ $state _state>](&self) -> bool {
+        for peer in self.instance_list.iter() {
+            // All instance must be at the desired state
+            if !peer.protocol.[< is_ $state _state>]() {
+                return false;
+            }
+        }
+        true
+    }
+}};
+}
+
+impl<P: Protocol + Send> Russula<P> {
+    state_api!(
+        /// Successfully connected to peer instances and ready to make progress.
+        ready
+    );
+    state_api!(
+        /// The coordination completed and reached a terminal state.
+        done
+    );
+    state_api!(
+        /// Note: Should only be called by Coordinators
+        ///
+        /// A intermediate state to query if Worker peers are in some
+        /// desired state.
+        worker_running
+    );
+}
+
+type SockProtocol<P> = (SocketAddr, P);
+pub struct RussulaBuilder<P: Protocol> {
+    /// Address on which the Coordinator and Worker communicate on.
+    ///
+    /// The Coordinator gets a list of workers addrs to 'connect' to. This can
+    /// be of size >= 1. The Worker gets its own addr to 'listen' on and should
+    /// be size = 1.
+    // TODO Create different Russula struct for Coordinator/Workers to capture
+    // different usage patterns.
+    protocol_addr_pair_list: Vec<SockProtocol<P>>,
+    poll_delay: Duration,
+}
+
+impl<P: Protocol> RussulaBuilder<P> {
+    pub fn new(peer_addr: BTreeSet<SocketAddr>, protocol: P, poll_delay: Duration) -> Self {
+        // TODO if worker check that the list is len 1 and points to local addr on which to listen
+        let mut peer_list = Vec::new();
+        peer_addr.into_iter().for_each(|addr| {
+            peer_list.push((addr, protocol.clone()));
+        });
+        Self {
+            protocol_addr_pair_list: peer_list,
+            poll_delay,
+        }
+    }
+
+    pub async fn build(self) -> RussulaResult<Russula<P>> {
+        let mut stream_protocol_list = Vec::new();
+        for (addr, protocol) in self.protocol_addr_pair_list.into_iter() {
+            let mut retry_attempts = 10;
+            loop {
+                if retry_attempts == 0 {
+                    return Err(RussulaError::NetworkConnectionRefused {
+                        dbg: "Failed to connect to peer".to_string(),
+                    });
+                }
+                match protocol.pair_peer(&addr).await {
+                    Ok(connect) => {
+                        info!("Coordinator: successfully connected to {}", addr);
+                        stream_protocol_list.push(ProtocolInstance {
+                            addr,
+                            stream: connect,
+                            protocol,
+                        });
+
+                        break;
+                    }
+                    Err(err) => {
+                        error!(
+                            "Failed to connect.. wait and retry. Try disabling VPN and check your network connectivity.
```

Comment:
Nit: If it failed 10 times, do we actually expect "wait and retry" to be a valid strategy? 😛

Reply:
A Coordinator attempts to connect to multiple workers. If the Coordinator is launched after the Worker, then retry can catch the failure.

Depending on the `poll_delay` the user passes in, the retries will run fast or slow. In practice I dont expect us to get to 10 very often.

Reply:
Figuring out that the error was happening due to VPN was not obvious so I dont think it hurts to leave it in there. If the user has VPN disabled then they can explore other solutions. I can clarify that disabling VPN could be one solution.